doc/reference/trd-userspace-readable-allow-syscalls.md

```diff
@@ -0,0 +1,188 @@
+
+| Syscall Class    | Syscall Class Number |
+|------------------|----------------------|
+| Read-Write Allow |           7          |
+
+The standard access model for userspace readable allowed buffers is that userspace can read
+from a buffer while the kernel can read or write. Syncronisation
+methods are required to ensure data consistency but are implementation specific.
+
+Simultaneous accesses to a buffer from both userspace and the kernel can
+cause userspace to read inconsistent data if not implemented properly.
+For example a userspace app could read partially written data.
+This would result in obscure timing bugs that are hard to detect. Due to this
+each capsule using the userspace readable allow mechanism MUST document, in a Draft or
+Final Documentary TRD, how it ensures userspace always reads consistent data
+from a userspace readable buffer.
+
+Finally, because a process conceptually relinquishes write access to a buffer
+when it makes a userspace readable allow call with it, a userspace API MUST NOT
+assume or rely on a process writing an allowed buffer. If userspace needs to
+write to a buffer held by the kernel, it MUST first regain access to it by
+calling the corresponding Userspace Readable Allow. A userspace API MAY
+allow a process to read an allowed buffer, but if it does, it must document
+a consistency mechanism.
+
+One example approach to ensure that userspace reads of a data object are
+consistent is to use a monotonic counter. Every time the kernel writes
+the data object, it increments the counter. If userspace reads the counter,
+reads the data object, then reads the counter again to check that it has not
+changed, it can check that the object was not modified mid-read. If the counter
+changes, it restarts the read of the data object. This approach is simple, but
+does make reading the data object take variable time and is theoretically
+vulnerable to starvation.
+
+An example of reading a monotonic counter from userspace would look like this:
+
+```c
+  volatile uint32_t high, low;
+
+  do {
+    // Read the current counter value
+    counter = ptr[0];
+
+    // Read in the data
+    my_data0 = ptr[1];
+    my_data1 = ptr[2];
+
+    // Only exit the loop if counter and ptr[0] are the same
+  } while (counter != ptr[0]);
+```
+
+where the counter is incremented on every context switch to userspace.
+
+3 libtock-c Userspace Library Methods
+=====================================
+
+3.1 Userspace Readable Allow
+---------------------------------
+
+The userspace readable allow system call class is how a userspace process
+shares a buffer with the kernel that the kernel can read and write.
+
+The userspace readable allow system call has this function prototype:
+
+```c
+typedef struct {
+  bool success;
+  void* ptr;
+  size_t size;
+  tock_error_t error;
+} userspace_readable_allow_return_t;
+
+userspace_readable_allow_return_t allow_userspace_readable(uint32_t driver, uint32_t allow, void* ptr, size_t size);
```

Comment:
```suggestion
userspace_readable_allow_return_t allow_userspace_readable(uint32_t driver, uint32_t allow, volatile void* ptr, size_t size);
```
Should this be marked `volatile` here? I think this buffer, once allowed, will always have C volatile semantics.

Somewhat oddly, the `ptr` returned in the `userspace_readable_allow_return_t`  will not be volatile, as that `ptr` will not have been / no longer shared with the kernel...